codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamChannel.java

```diff
@@ -0,0 +1,304 @@
+                pipeline().fireChannelReadComplete();
+            }
+        }
+    };
+
+    private boolean closed;
+    private boolean readInProgress;
+
+    public AbstractHttp2StreamChannel(Channel parent) {
+        super(parent);
+    }
+
+    @Override
+    public ChannelMetadata metadata() {
+        return METADATA;
+    }
+
+    @Override
+    public ChannelConfig config() {
+        return config;
+    }
+
+    @Override
+    public boolean isOpen() {
+        return !closed;
+    }
+
+    @Override
+    public boolean isActive() {
+        return !closed;
+    }
+
+    @Override
+    protected AbstractUnsafe newUnsafe() {
+        return new Unsafe();
+    }
+
+    @Override
+    protected boolean isCompatible(EventLoop loop) {
+        return true;
+    }
+
+    @Override
+    protected SocketAddress localAddress0() {
+        return parent().localAddress();
+    }
+
+    @Override
+    protected SocketAddress remoteAddress0() {
+        return parent().remoteAddress();
+    }
+
+    @Override
+    protected void doBind(SocketAddress localAddress) throws Exception { }
+
+    @Override
+    protected void doDisconnect() throws Exception {
+        closed = true;
+    }
+
+    @Override
+    protected void doClose() throws Exception {
+        closed = true;
+        while (!inboundBuffer.isEmpty()) {
+            ReferenceCountUtil.release(inboundBuffer.poll());
+        }
+    }
+
+    @Override
+    protected void doBeginRead() {
+        if (readInProgress) {
+            return;
+        }
+        unsafe().recvBufAllocHandle().reset(config());
+
+        if (inboundBuffer.isEmpty()) {
+            readInProgress = true;
+            return;
+        }
+
+        final RecvByteBufAllocator.Handle allocHandle = unsafe().recvBufAllocHandle();
+        do {
+          Object m = inboundBuffer.poll();
+          if (m == null) {
+            break;
+          }
+          if (!doRead0(m, allocHandle)) {
+            return;
+          }
+        } while (allocHandle.continueReading());
+
+        allocHandle.readComplete();
+        pipeline().fireChannelReadComplete();
+    }
+
+    @Override
+    protected final void doWrite(ChannelOutboundBuffer in) throws Exception {
+        if (closed) {
+            throw new ClosedChannelException();
+        }
+
+        EventExecutor preferredExecutor = preferredEventExecutor();
+
+        if (preferredExecutor.inEventLoop()) {
+            for (;;) {
+                Object msg = in.current();
+                if (msg == null) {
+                    break;
+                }
+                try {
+                    doWrite(ReferenceCountUtil.retain(msg));
+                } catch (Throwable t) {
+                    // It would be nice to fail the future, but we can't do that if not on the event
+                    // loop. So we instead opt for a solution that is consistent.
+                    pipeline().fireExceptionCaught(t);
```

Comment:
In my mind, not really. Even with your suggestion, the behavior will still be very broken. We can't use it until this is actually, really, truly solved. Trying to fix a case or two won't make it any more production-ready. We need to solve the root problem, which relates to `ChannelOutboundBuffer`, not event loops. I fully expect this will become easy once we have outbound flow control and that any workarounds we develop now won't aid us at that point. So not wasting effort here until then seems quite reasonable.